### PR TITLE
Add a built-in lens correction module.

### PR DESCRIPTION
**Pull request by FreddieWitherden**
This draft PR adds support for a manufacturer lens correction module which can correct distortion, chromatic aberration, and vignetting using the lens correction coefficients stored in the RAW file by the manufacturer.  Currently Sony and Fuji APS-C cameras are supported.

Note that in order to test the module it is necessary to start Darktable with `--library :memory:` and one must also be using the absolute latest Git revision of Exiv 0.27.  The library flag is require since this PR must pull some additional data out of the RAW file into `dt_image_t`.  However, `dt_image_t` appears to be cached in a SQL database which will require changes in order to serialise the new structure.

### CODE REVIEW

**Comment by johnny-bit** [comment]
There are couple of things you might want to consider:

- our build system allows module to be conditionally built based on presence of library the module depends on, for example you might want to check out lut3d
- given that module only works for certain sony & fuji cameras with certain data, you might want to disable/hide completely module based on what image is being processed, for example you can check (i think) rotate pixels and white balance modules on that :)
- a tiny code nitpick - you don't need to include color-picker-proxy as your module doesn't include color pickers at all :)

**Comment by FreddieWitherden** [comment]
Thank you for your feedback.

>  our build system allows module to be conditionally built based on presence of library the module depends on, for example you might want to check out lut3d

One nice aspect of this module is that it is dependency free.  All correction is performed by mlens.c.

> given that module only works for certain sony & fuji cameras with certain data, you might want to disable/hide completely module based on what image is being processed, for example you can check (i think) rotate pixels and white balance modules on that :)

I think I handle this with the check in `reload_defaults` which if `img->exif_correction_type` is 0 sets `module->hide_enable_button`.  Let me know if there is a preferred means, however.

>  a tiny code nitpick - you don't need to include color-picker-proxy as your module doesn't include color pickers at all :)

Fixed.

**Comment by johnny-bit** [comment]
> One nice aspect of this module is that it is dependency free.

Right, I meant that it's working depends on having exiv2 v 0.27.4 (or dev branch?) but it's not actually IN module, but outside of it... But if i see correctly you already handle that case...

> Let me know if there is a preferred means, however.

uh oh... @TurboGit will surely guide that ;)

**Comment by johnny-bit** [comment]
2 more things of which i forgot ;)

It would be awesome if you could add unit tests according to https://github.com/darktable-org/darktable/tree/master/src/tests/unittests :)
And it might be worth considering having also integration tests for results once you're certain that the algorithms are stable according to https://github.com/darktable-org/darktable/tree/master/src/tests/integration

**Comment by TurboGit** [comment]
> I think I handle this with the check in `reload_defaults` which if `img->exif_correction_type` is 0 sets `module->hide_enable_button`. Let me know if there is a preferred means, however.

There is already some modules doing that, for example: rawdenoise or cacorrect which are only RAW pictures. And all this is happening in reload_defaults() indeed and plays with the module->hide_enable_button plus a message on the iop to explain why it is disabled.

**Comment by FreddieWitherden** [comment]
The relevant patches have now been merged into the 0.27-maintenance branch of Exiv2 so this should be ready for testing.

**Comment by aurelienpierre** [comment]
I have mixed feelings about that:

1. why not add that to the current lens correction module ? Having 2 modules for the same task is not great.
2. the interpolation is not really an interpolation, it only computes the coordinates of the nearest neighbour, if I'm not mistaken, and that will not produce good results for any kind of sharp pattern (fences, grids, etc.)

**Comment by FreddieWitherden** [comment]
>  1. why not add that to the current lens correction module ? Having 2 modules for the same task is not great.

The existing module has a dependency on an external library (lensfun) whereas this one does not.

> 2. the interpolation is not really an interpolation, it only computes the coordinates of the nearest neighbour, if I'm not mistaken, and that will not produce good results for any kind of sharp pattern (fences, grids, etc.)

I'm not entirely sure what you mean here.  In the module there are two types of interpolation: model coefficient interpolation which is accomplished using a linear spline (see `interpolate`), and actual pixel data interpolation which is performed using `dt_interpolation_compute_sample`.

**Comment by aurelienpierre** [comment]
> The existing module has a dependency on an external library (lensfun) whereas this one does not.

I understand, but workflow-wise, having 2 different modules to perform the same task is weird. Also, users have asked a way to tune the Lensfun profiles for distortion and vignetting, which this PR does, but separately. It would be nicer to blend this feature in the `lens.cc` module as an additional path, use macros to enable the Lensfun path if the lib is detected, and otherwise just offer the option developed here. Then add user corrections over the Lensfun profiles too.

> I'm not entirely sure what you mean here. In the module there are two types of interpolation: model coefficient interpolation which is accomplished using a linear spline (see interpolate), and actual pixel data interpolation which is performed using dt_interpolation_compute_sample.

Ok that's good then.

**Comment by FreddieWitherden** [comment]
> > The existing module has a dependency on an external library (lensfun) whereas this one does not.
> 
> I understand, but workflow-wise, having 2 different modules to perform the same task is weird. Also, users have asked a way to tune the Lensfun profiles for distortion and vignetting, which this PR does, but separately. It would be nicer to blend this feature in the `lens.cc` module as an additional path, use macros to enable the Lensfun path if the lib is detected, and otherwise just offer the option developed here. Then add user corrections over the Lensfun profiles too.

Although the functionality is similar the means by which they accomplish it is somewhat different.  A good example of this is fine-tuning.  It is trivial in my module (which is spline based) as you simply tweak the values towards/away from unity.  This is all done as a pre-processing step with no overhead.  For lensfun, whose models are monomial coefficient based, this is simply not possible as the model response is not linear to changes in the coefficients.  Thus one would need to do corrections on a per-pixel basis, computing the displacement vector and rescaling appropriately.  This is not only messy but also expensive as you're piling on additional square root operations.  The result is two quite different implementations.

Similarly, for OpenCL things are different.  For lensfun you need to compute the distorted coordinates by calling lensfun methods on the CPU in one big array and then pass it into the kernel.  This is slow, both due to the computational cost of the operations and the time required for the copy.  In contrast, my module can be completely offloaded.

**Comment by aurelienpierre** [comment]
> Although the functionality is similar the means by which they accomplish it is somewhat different.

I know that. 

Right now, in `lens.cc`, you have one big `process()` function. What I mean is, rename that function `process_lensfun()`, and include the `process()` of the current PR in there too, renamed as a `process_builtin()`. Then, replace the `lens.cc` `process()` with:
```
void process(args)
{
  if(method == DT_LENS_BUILT_IN)
    process_builtin(args);
#ifdef LENSFUN
  else if(method == DT_LENS_LENSFUN);
    process_lensfun(args)
#endif
   else
     // some fall-back or error log
}
```
and add a `method` parameter taking values into a
```
typedef enum dt_lens_method
{
   DT_LENS_BUILT_IN = 0,
   DT_LENS_LENSFUN = 1
} dt_lens_method;
```
that can be set through a combobox, by the user, in GUI.

That way, we can merge the feature in the same module, while still having different code pathes. Add the parameters you need in the structure, and split other functions like `distort_transform()` with the same logic.

**Comment by TurboGit** [comment]
I would really like to see Aurélien's proposal being tested/implemented. I seems a good candidate avoid code duplication, BUT... beware that in such module we need all the transforms routines:

- distort_transform
- distort_backtransform
- distort_mask

Those are mandatory to have the masks and other modules to work properly when having the transformation activated.

**Review by TurboGit — CHANGES_REQUESTED** [review]
try to integrate this into current lens.cc
implements the transform routines

**Comment by FreddieWitherden** [comment]
I believe that the two distort transform routines are already implemented, and I can look into implementing the mask function.  However, my big issue is around the changes to `dt_image_t`, which seem to have a negative interaction with the SQL based caching mechanisms in Darktable.  This is code I am unfamiliar with and I am hesitant to make any changes to the DB side of things.

**Comment by TurboGit** [comment]
> I believe that the two distort transform routines are already implemented

I see, sorry I missed that.

**Comment by TurboGit** [comment]
> However, my big issue is around the changes to `dt_image_t`, which seem to have a negative interaction with the SQL based caching mechanisms in Darktable.

Not sure to see your problem. But if we can help when you'll have detailed report about the issue when implemented.

**Comment by aurelienpierre** [comment]
@FreddieWitherden don't hesitate to ask if you need help or explanations about dt's guts.

**Comment by TurboGit** [comment]
@FreddieWitherden : Just checking if this PR is still on your radar and if you are ok to integrate that into the current lens module. If you need help please let us know. TIA.

**Comment by FreddieWitherden** [comment]
It is still on my radar, but I do not think I'll have time to revisit it until the summer I'm afraid.

**Comment by aurelienpierre** [comment]
Otherwise, we can merge with current module if you want. The feature is cool, I would like it ASAP in master. Unless you want to do it yourself, which would be understandable.

**Comment by FreddieWitherden** [comment]
> Otherwise, we can merge with current module if you want. The feature is cool, I would like it ASAP in master. Unless you want to do it yourself, which would be understandable.

No problem at all.

**Comment by janfri** [comment]
> > Otherwise, we can merge with current module if you want. The feature is cool, I would like it ASAP in master. Unless you want to do it yourself, which would be understandable.
> 
> No problem at all, please go ahead.

Is there any progress? If I understand @FreddieWitherden correct, he means the current implementation could be merged, as @aurelienpierre suggested.

**Comment by aurelienpierre** [comment]
What I meant was to include this PR as another option in the current lens module ("merging with current module"), to avoid having 2 lens modules which would only confuse users (especially since they would not be compatible with each other). Not merging this PR as-is.

**Comment by janfri** [comment]
> It is still on my radar, but I do not think I'll have time to revisit it until the summer I'm afraid.

@FreddieWitherden Sorry for my impatience, but it would be so great to have this feature in darktable. Could we support you in some way? (The summer is coming to the end on the northern hemisphere ... ;-)

**Comment by TurboGit** [comment]
No answer since 10 months so I'll close. If this is pursued let me know and I'll reopen. Thanks.

**Comment by FreddieWitherden** [comment]
So just an update on where I got to over the summer.  I started working to merge this module with the existing lens correction module in lens.cc.  The code in this module is not only C++ but exceptionally low quality C++ at that.  Case in point it comes in at ~2500 lines whereas mlens.c — which does all of the correction itself rather than calling out to a third party library — comes in at just ~450 lines.  Part of this is because the lensfun API is utterly atrocious and not fit for purpose.

Additionally, the models used by lensfun are based around monomials in R.  It is difficult to articulate how silly this is.  The result is models whose parameters are twitchy and difficult to interpret; two corrections which are very similar can have very different coefficients.  The fact these models are not sensible shows itself very clearly when lensfun tries to interpolate between corrections (say you have one at 70mm, another at 90mm but your photograph is taken at 80mm).  This combined with the fact that lensfun is moribund is reason enough to drop it.

So, I started migrating the lensfun correction database to spline based models (which is what Fuji and Sony cameras use) and also from XML to JSON.  The idea being to have something which Darktable can use directly, eliminating the dependency on lensfun.  Moreover, this is also likely to be substantially more efficient especially when OpenCL is used (as the model can be evaluated in the kernel rather than having lensfun evaluate it at all sample points into a buffer which then has to be uploaded).  Further, down the line this opens up the possibility for users to create their own corrections in Darktable (most of the algorithms/code already exist in models in one form or another).

I made a little bit of progress in this area over the course of a couple of weekends, although can not for the life of me find the code.  Either way, I am still convinced this is the best path to take.  It eliminates a dependency, cuts down on the amount of C++ code, and enables corrections to find their way into the hands of users quicker (since the JSON corrections can be shipped with Darktable).  I am one again tight for time, but would strongly suggest anyone who is considering working on the existing lens correction module focus on this approach.

**Comment by TurboGit** [comment]
Sounds promising, thanks for the update. Let me reopen. What I'm not clear yet is how is the data model for lens are constructed? Are you saying that you take the lensfun database and convert it to JSON? And from there you use your new module to do the correction? If not, can you describe this part a bit more?

Also, are you taking about all corrections made by lensfun (geometric, chromatic abberations...)?

TIA.

**Comment by cryptomilk** [comment]
I could add json output to [lens_calibrate](https://gitlab.com/cryptomilk/lens_calibrate/) if needed.

**Comment by FreddieWitherden** [comment]
> Sounds promising, thanks for the update. Let me reopen. What I'm not clear yet is how is the data model for lens are constructed? Are you saying that you take the lensfun database and convert it to JSON? And from there you use your new module to do the correction? If not, can you describe this part a bit more?
> 
> Also, are you taking about all corrections made by lensfun (geometric, chromatic abberations...)?

Constructing the data model is relatively simple.  Lens aberrations are quantified as a function of the radius R from the centre.  In the spline model used by my module we pick some number of points between R = 0 and R = 1 (normalised) and state the amount of aberration at this point (distortion, vignetting, CA R, CA, B).  Lensfun has a range of different models for aberrations but they all result in a function f(R).  Thus the conversion is a simple matter of picking a number of points in [0,1] and evaluating f(R) at these points.  Getting f(R) requires a little work due to some of the silly games lensfun plays with coefficients but it isn't too hard.

I accomplished this using a Python script which I believe was ~100 lines or so which read in the XML and outputted JSON (although at the time I stopped work on this I was still fiddling around with the precise format of the JSON).  Spline models are also easier to create corrections for as you simply need to know the amount of aberration at a specific point.  In contrast the lensfun models typically require one to form and solve a linear least squares problem to obtain the final coefficients.

**Comment by TurboGit** [comment]
Thanks for the clarification. My point was to understand what would it cost to move to the new module. I mean, even if the math are better if there is no model so no lens supported it won't be good and rebuilding the models for all lensfun supported camera/lens will be a huge effort. So, we really need to have a way to migrate the lensfun database to new module and ensure that we can import new lens from this same database from time to time too.

**Comment by TurboGit** [comment]
> I could add json output to [lens_calibrate](https://gitlab.com/cryptomilk/lens_calibrate/) if needed.

Probably a good option, yes!

**Comment by FreddieWitherden** [comment]
> Thanks for the clarification. My point was to understand what would it cost to move to the new module. I mean, even if the math are better if there is no model so no lens supported it won't be good and rebuilding the models for all lensfun supported camera/lens will be a huge effort. So, we really need to have a way to migrate the lensfun database to new module and ensure that we can import new lens from this same database from time to time too.

Right.  The way I envisioned this working was before each major release of Darktable one would run the Python script to hoover up the current lensfun database from GitHub and produce an easier to use JSON database which can be shipped as part of Darktable (easier as there is a single unified model, and the JSON is slightly easier to work with in C than XML).  Should also resolve user-complaints around a lack of regular releases of lensfun and the need for them to periodically run the database upgrade function.

**Comment by Fabi2607** [comment]
If you don't have time to work at 1. and 2. At the moment, I would offer to take a look. It looks like we could also use the blob type to store the information in the database, but I am unsure if we should go this route. It looks like a nightmare in terms of future extensibility.

Maybe @aurelienpierre or @TurboGit could give a few pointers for the best path forward?

**Comment by Fabi2607** [comment]
I do have the basic caching code working but I am still missing two vital parts:

1. What's the best way of handling existing images in the database? The columns are not filled and we would need to do this after migrating the database schema
2. If we extend the data structure with more makers, the current implementation would have issues handling this. AFAIC we could avoid this by only storing the maker specific correction data and only reading that one back from the database?

**Comment by FreddieWitherden** [comment]
> I do have the basic caching code working but I am still missing two vital parts:
> 
>     1. What's the best way of handling existing images in the database? The columns are not filled and we would need to do this after migrating the database schema
> 
>     2. If we extend the data structure with more makers, the current implementation would have issues handling this. AFAICS we could avoid this by only storing the maker specific correction data and only reading that one back from the database?

I would try storing the data as JSON which would enable you to use a TEXT field.  This would allow the column to be extensible going forwards.

**Comment by TurboGit** [comment]
Not sure to be of any help at this stage and so close of a release.

There is one point that is not clear to me. What are the data you are retrieving during the import (modification of  dt_image_loader_t)? Why you cannot cache this in SQL later? At the moment many information read at download time are added into the database, so my question is why not possible in this case?

Can you describe the data in question?

If an array as you seems to indicate we may need another table:
`
   imgid | key | value`

Which is the way to handle "arrays" in SQL. Anyway we really need more information. TIA.

**Comment by Fabi2607** [comment]
Hi,

To clarify a bit:

The information being read during import are fixed length arrays describing the parameters for length correction, depending on the manufacturer. 

My current approach would be:
Store the data type as an enum (Integer) and store the maker specific data as a blob. If we need to extend this, we can extend the enum. Same if the format for a manufacturer changes with new cameras. This would simply become a new lens correction type.

My (only) question that currently remains is how already imported images are handled if fields are added. Is there a mechanism in darktable to read the exif data again and fill the fields?

**Comment by Fabi2607** [comment]
@FreddieWitherden I took a few pictures from rawsamples.ch and the lens correction for Fuji seems to be different:

```
RAW_FUIJI_X-E2.RAF    Exif.Fujifilm.GeometricDistortionParams      SRational  23  2942/11 0/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 0/1000 29/1000 72/1000 142/1000 251/1000 410/1000 616/1000 859/1000 1173/1000 1553/1000 2021/1000
RAW_FUIJI_X-E2.RAF    Exif.Fujifilm.ChromaticAberrationParams      SRational  31  2942/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 140/1000000 257/1000000 326/1000000 325/1000000 274/1000000 221/1000000 156/1000000 117/1000000 16/1000000 -91/1000000 -91/1000000 -166/1000000
 -209/1000000 -204/1000000 -152/1000000 -72/1000000 23/1000000 132/1000000 257/1000000 396/1000000
RAW_FUIJI_X-E2.RAF    Exif.Fujifilm.VignettingParams               SRational  23  2942/11 0/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100 10000/100
RAW_FUJI_F600XR.RAF   Exif.Fujifilm.GeometricDistortionParams      SRational  23  2030/11 0/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000 0/1000
RAW_FUJI_F600XR.RAF   Exif.Fujifilm.ChromaticAberrationParams      SRational  31  2030/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 381/1000000 992/1000000 992/1000000
 992/1000000 992/1000000 992/1000000 992/1000000 992/1000000 992/1000000 992/1000000 992/1000000
RAW_FUJI_F600XR.RAF   Exif.Fujifilm.VignettingParams               SRational  23  2030/11 0/10 1/10 2/10 3/10 4/10 5/10 6/10 7/10 8/10 9/10 10/10 10000/100 9999/100 9985/100 9928/100 9778/100 9475/100 8971/100 8247/100 7339/100 6327/100 5305/100
```

**Comment by FreddieWitherden** [comment]
The coefficients need processing as:

https://github.com/FreddieWitherden/darktable/blob/master/src/iop/mlens.c#L90

In terms of the Sony coefficients, for reasons I never got to the bottom of Sony embeds several sets of correction coefficients into their images, often with slight differences.  The ones which are currently pulled out by the module are the same as those used by Imaging Edge.

**Comment by Fabi2607** [comment]
> The coefficients need processing as:
> 
> https://github.com/FreddieWitherden/darktable/blob/master/src/iop/mlens.c#L90
> 
> In terms of the Sony coefficients, for reasons I never got to the bottom of Sony embeds several sets of correction coefficients into their images, often with slight differences. The ones which are currently pulled out by the module are the same as those used by Imaging Edge.

I saw that code, what I am saying is that for Sony, older cameras store the parameters in "Exif.Sony2010e". And the assumption `len(VignettingCorrParams) == 2*len(ChromaticAberrationCorrParams) == len(DistortionCorrParams)` is wrong. Coded as `nc <= 16 && 2*nc == posc->toLong(0) && nc == posv->toLong(0)` is wrong.

For Fuji the assumption `(posd->count() == 19 && posc->count() == 29 && posv->count() == 19)` is wrong.

 I think we should treat distortion, vignetting and chromatic aberration as independent.

File: src/iop/mlens.c [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+{
+  return IOP_FLAGS_ALLOW_TILING | IOP_FLAGS_TILING_FULL_ROI | IOP_FLAGS_ONE_INSTANCE | IOP_FLAGS_UNSAFE_COPY;
+}
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+static int init_coeffs(const dt_image_t *img, const dt_iop_mlens_params_t *d,
```
Comment (Fabi2607) [inline src/iop/mlens.c:90]: Looking at this in more detail:
Why don't we do this processing on import and store a manufacturer independent matrix for distortion, ca and vignetting? That would simplify the code here, keep dt_image_t small and simply the caching code.

**Comment by FreddieWitherden** [comment]
> > The coefficients need processing as:
> > https://github.com/FreddieWitherden/darktable/blob/master/src/iop/mlens.c#L90
> > In terms of the Sony coefficients, for reasons I never got to the bottom of Sony embeds several sets of correction coefficients into their images, often with slight differences. The ones which are currently pulled out by the module are the same as those used by Imaging Edge.
> 
> I saw that code, what I am saying is that for Sony, older cameras store the parameters in "Exif.Sony2010e". And the assumption `len(VignettingCorrParams) == 2*len(ChromaticAberrationCorrParams) == len(DistortionCorrParams)`, coded as `nc <= 16 && 2*nc == posc->toLong(0) && nc == posv->toLong(0)` is wrong.
> 
> For Fuji the assumption `(posd->count() == 19 && posc->count() == 29 && posv->count() == 19)` is wrong.
> 
> I think we should treat distortion, vignetting and chromatic aberration as independent.

I see what you mean.  However, I do not think it is as simple as just treating them as being independent; it is possible that those files use completely different correction models to those of current cameras.  This is why the checks in the current code are quite strict in terms of what they will accept.  Someone with such a camera will need to take a series of test shots and then verify exactly how the corrections are encoded.

File: src/iop/mlens.c (hunk 1 above)
Comment (FreddieWitherden) [reply src/iop/mlens.c:90]: As the models are reverse engineered there is no guarantee that the coefficient processing is correct (I think it is, but I am not sure).  Hence why I placed it in a function.  Incidentally, this is why I am against the idea of caching these coefficients in any manner at all.  It is almost certainly more trouble than it's worth.  When you're opening a 20 MiB raw file and doing highly intensive processing on it, shaving off a few milliseconds by caching metadata is basically a rounding error.  Moreover, it leads to issues such as those you've outlined above regarding how to deal with files whose metadata was extracted before support was added.

Comment (Fabi2607) [reply src/iop/mlens.c:90]: Fair point.

**Comment by cryptomilk** [comment]
The best would be if there would be an independent project for the lens data. So it can be used by other projects like vkdt, filmulator or rawtherapee. Darktable should just try to update the database on startup.

The data could also be consumed by lensfun if they want. currently it can take several years to get a pull request merged ...

**Comment by cryptomilk** [comment]
Well maybe we can come up with a database projects which has data for rawspeed, exiv2, lensfun, darktable etc. So that the basic metadata about a camera or lens comes from a single project.

**Comment by FreddieWitherden** [comment]
> Well maybe we can come up with a database projects which has data for rawspeed, exiv2, lensfun, darktable etc. So that the basic metadata about a camera or lens comes from a single project.

Maybe.  I have never been quite sure why Lensfun needs to know about specific camera models.  All it should need to know is the lens mount and crop factor.

In terms of the lens data itself, virtually all new first-party lenses (and some third party lenses) for all new mounts embed correction data in the RAW files.  Hence, if we can reverse engineer these formats (I've done and coded Fuji X and Sony E) then there should be much less of a need to update the data since new lenses will just work automatically.

For lenses which do require explicit data I would much rather it come with Darktable in the same way that rawspeed does rather than being a standalone library such as lensfun.

**Comment by gunarka** [comment]
I had a hard time trying to profile my lens for lensfun. So, this is a great module - many thanks!
How can also Nikon Z be integrated?

**Comment by FreddieWitherden** [comment]
> How can also Nikon Z be integrated?

This will require reverse engineering how correction data is embedded in the Nikon RAW files.  For Fuji and Sony the reverse engineering was relatively simple as exiftool already knew where in the files the correction data was located so all I had to do was interpret how the correction data is used.  For Nikon Z I suspect some more work well be needed, although there is very much a methodology to it.

**Comment by greenoid** [comment]
Exiftool knows about the correction parameters in Panasonic RW2 rawfiles.
I try to teach exiv2 the same, so your module may correct them to.

**Comment by greenoid** [comment]
I added panasonic makernotes for distortion correction and chromatic aberration data to exiv2 as a PR:
https://github.com/Exiv2/exiv2/pull/2077
Hope it gets accepted.

**Comment by uchrisu** [comment]
@FreddieWitherden can you try 
`knots[i] = (float) (i+0.5) / (nc - 1);`
and maybe adding a point (0,0)

instead of 

`knots[i] = (float) i / (nc - 1);`

With i/(nc-1) the fitting does not seem to be perfect, but with (i+1)/(nc-1) I got nearly perfect fittings.

**Comment by FreddieWitherden** [comment]
> @FreddieWitherden can you try `knots[i] = (float) (i+0.5) / (nc - 1);` and maybe adding a point (0,0)
> 
> instead of
> 
> `knots[i] = (float) i / (nc - 1);`
> 
> With i/(nc-1) the fitting does not seem to be perfect, but with (i+1)/(nc-1) I got nearly perfect fittings.

This is entirely plausible.  I reverse engineered the model and was never entirely sure if there should be a knot at 0 or not.  I am somewhat less sure about if a `+ 0.5` is needed; from a model standpoint this would imply the final knot is slightly outside of the frame.  A good test for this would be on an APS-C crop image (which have fewer knots and so the impact of `+ 0.5` is relatively larger due to the smaller denominator).

**Comment by uchrisu** [comment]
I checked it by plotting the green Value over relative radius (corner is 1) along a line from the center to the corner of the image.
The comparison is between the distorted (raw) image and an image with in-camera correction distorted by the coefficients (linear interpolation) afterwards. In the case of correct interpretation of the coefficients the newly distorted image with in-camera correction should look the same as the distorted image (after an appropriate scaling)

With (i+0.5)/(nc-1) I got:


Now, here are some crops from 3 approaches:
i/(nc-1):
crop2_15

(i+0.5)/(nc-1):
crop2_15_5

(i+1)/(nc): 
crop2_16

The last one was my first approach and gives good, but not perfect, results. The deviations of the middle one are withing in the range of one pixel. The fit seems to be perfect. The notches are perfectly aligned.

**Comment by FreddieWitherden** [comment]
I'm convinced.  Can you confirm it also works for an APS-C crop?  Most full frame cameras can be asked to produce these files.

**Comment by uchrisu** [comment]
In APS-C it is little bit more complicated. I have spotted that the 11 coefficients are exactly the same as the first 11 of the 16 full frame coefficients. However, the crop factor is not exactly 1.5 but rather something like 1.524 (in case of Sony A7III). Thus the knots should be scaled by the factor (15/10 / crop) to get the best fitting, when the corner of the APS-C image should have the radius 1. With this scaling, the knots are located at the same pixels of the (full-frame) sensor in APS-C mode compared to the full-frame mode - as it should be, since we don't change the optical system. 

Overview:
apsc_11_full

Now zoomed in with knot[i] = (i+0.5)/(nc -1) = (i+0.5)/10:
aps_11_crop

And zoomed in with knot[i] =  (i+0.5)/(nc -1) * (15/10/real_crop) = (i+0.5)/10 * (1.5/1.524):
aps_fromfull_crop

However, also the fitting without the crop-correction factor is quite good.

**Comment by FreddieWitherden** [comment]
I may have some time over the coming months to revisit this PR, or at least get it building against the current release.

My current open query is how, in the module init function, can I get access to the EXIF data?  This would avoid needing to change the DB structure to include the extracted correction data - which is non-trivial and not robust against updates to Exiv2 (which may include support for previously unsupported metadata over time, rendering Darktable's cache stale).

**Comment by paolodepetrillo** [comment]
@FreddieWitherden Check out the recent changes that were made in this commit https://github.com/darktable-org/darktable/commit/9552de237e70b6811214950bc8c30b6f9cce3d1c to load some large variable sized data from exif tags into memory without making any changes to the database - it could easily be extended to also load the lens correction data the same way.

**Comment by FreddieWitherden** [comment]
> @FreddieWitherden Check out the recent changes that were made in this commit [9552de2](https://github.com/darktable-org/darktable/commit/9552de237e70b6811214950bc8c30b6f9cce3d1c) to load some large variable sized data from exif tags into memory without making any changes to the database - it could easily be extended to also load the lens correction data the same way.

Done. I've also rebased against the newest version of dt and applied the Sony coefficient fix suggested by @uchrisu.

**Comment by sgotti** [comment]
I'm using this modules for some time (with Fuji cameras and Fuji and Tamron lenses) and it works really well! Great work @FreddieWitherden!

The latest version that reads exit data on file open without updating the DB fixed different issues. 
(Probably the PR needs some cleanup since there're some unwanted changes with submodules).

Additionally the module logic/code has been also used inside ART by @agriggio and it's used by default when the lens correction metadata is available.

**Comment by FreddieWitherden** [comment]
> The latest version that reads exif data on file open without updating the DB fixed different issues. (Probably the PR needs some cleanup since there're some unwanted changes with submodules).

Yes, I think that the module is in relatively good shape.  It would be nice if it could get merged (in one form or another) since some of the lenses Sony and Fuji are bringing out have *a lot* of distortion.

**Comment by TurboGit** [comment]
The module maybe, but the PR is not really in good shape. You need to rebase it and avoid all modifications on submodules (LibRAW, OpenCL, integration...).

A second module for lens distortion correction is not really clean. But I may understand that merging it in the current lens correction module may be difficult. That being said having a built-in mode in the current lens correction module will be better I think. We could for example add a mode "auto" (or build-in) in the current module when data are available from EXIF.

This has already been discussed BTW and I still think it would be the way to go.

**Comment by TurboGit** [comment]
@FreddieWitherden : Let me know if you plan on working on the redesign of the module here for integration into the current lens correction module. Otherwise we could just close, let me know. TIA.

**Comment by FreddieWitherden** [comment]
> @FreddieWitherden : Let me know if you plan on working on the redesign of the module here for integration into the current lens correction module. Otherwise we could just close, let me know. TIA.

I don't think I'll have the time to look into this in the near future I'm afraid.

**Comment by TurboGit** [comment]
> @TurboGit I'm finishing a PR that will merge @FreddieWitherden work inside the lens module.

That's a great news to see this project being taken, thanks!

**Comment by paolodepetrillo** [comment]
I had started working on adding support for Olympus (micro four thirds) to this PR, should be able to add it to @sgotti version.

I was able to figure out distortion and CA by comparing the exif tags in the raw file to the output of Adobe DNG Converter. Distortion and CA are encoded as a polynomial a + b*r^2 + c*r^4 + d*r^6, same as the WarpRectilinear DNG opcode, and can easily be converted to splines like Sony and Fuji use.

Not sure how vignetting works - these cameras have an option 'shading compensation' which applies the vignetting to the actual raw image contents, so I don't know if the same information is also included in a tag that could be used if the option wasn't enabled when the photo was taken.

**Comment by sgotti** [comment]
Created PR #12714

**Comment by sgotti** [comment]
> I had started working on adding support for Olympus (micro four thirds) to this PR, should be able to add it to @sgotti version.
> 
> I was able to figure out distortion and CA by comparing the exif tags in the raw file to the output of Adobe DNG Converter. Distortion and CA are encoded as a polynomial a + b_r^2 + c_r^4 + d*r^6, same as the WarpRectilinear DNG opcode, and can easily be converted to splines like Sony and Fuji use.
> 
> Not sure how vignetting works - these cameras have an option 'shading compensation' which applies the vignetting to the actual raw image contents, so I don't know if the same information is also included in a tag that could be used if the option wasn't enabled when the photo was taken.

@paolodepetrillo Great! Noticed that ART (that uses the @FreddieWitherden algorithm) added corrections based on DNG files metadata. Also this could be also added and perhaps that code can be of help also to you for olympus lens corrections.

**Comment by greenoid** [comment]
I need help with guessing the distortion formula of the embedded distortion correction values embedded in newer Panasonic Lumix S (full-frame) RW2 raw files. None of the correction formulars of lensfun does the job. Distortion correction parameters are embedded as an array of 16 integer values in the EXIF oem extension, position is know, if have example photos, don't know how the mathematics of anti distortion works.

**Comment by roth-m** [comment]
> I need help with guessing the distortion formula of the embedded distortion correction values embedded in newer Panasonic Lumix S (full-frame) RW2 raw files. None of the correction formulars of lensfun does the job. Distortion correction parameters are embedded as an array of 16 integer values in the EXIF oem extension, position is know, if have example photos, don't know how the mathematics of anti distortion works.

I have matched 2 WarpRectilinear coefficients out of the 4 in Tag 0x0119. I'm also investigating Tag 0x011B which contains much more coefficients but I could not correlate one of them with any DNG OpCodeList WarpRectilinear one.
I can provide you with tools to dump coefficients of these 2 EXIF Tags and coefficients from DNG OpCodeList3.
I'm currently using R on data with focal + Tags dumps + DNG dumps.

**Comment by roth-m** [comment]
> > I have matched 2 WarpRectilinear coefficients out of the 4 in Tag 0x0119.
> 
> As RT has borrowed from dt, [you could claim something back now](https://github.com/Beep6581/RawTherapee/pull/7100)? 😉 @roth-m
> 
> See also:
> 
> https://syscall.eu/archive.html https://www.andrewj.com//mft/algorithms.php

Unfortunately the references you mention provide approximative formulas (sometimes it is really bad) for Barrel distorsion.

DNG is based on Brown Conrady distorsion and I believe the data in Pana Exif have to be reversed. 
Indeed, the formulas in https://www.mdpi.com/1424-8220/16/6/807 work nearly perfectly for 1 channel (as you know CA is applied on DNGs coefficients).

I did a dump of DNG coefficients for 10000 pictures (same principle as it has been done for Olympus) and successfully fitted polynoms (order 3) on all coefficients using 3 to 5 coefficients in Exif. But not sure it would fit for all lenses.

**Pull request closed without merging**